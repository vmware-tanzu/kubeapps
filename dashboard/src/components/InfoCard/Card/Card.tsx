// Copyright 2020-2023 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import { HTMLAttributes } from "react";

export interface ICardProps {
  children: React.ReactNode;
  clickable?: boolean;
  onClick?: (e: React.MouseEvent<HTMLDivElement>) => void;
}

const Card = ({ children, clickable = false, onClick }: ICardProps) => {
  // Common props for the element
  const innerProps: HTMLAttributes<HTMLDivElement> = {
    className: `card ${clickable || typeof onClick === "function" ? "clickable" : ""}`,
    onClick: onClick,
  };

  // Make the card focusable by keyboard navigation. I'm not adding it based on the
  // clickable prop because I assume there's something around the card that manages
  // the onClick event.
  if (typeof onClick === "function") {
    innerProps.tabIndex = 0;

    // Runs onClick when the user types `enter` or `space`
    innerProps.onKeyDown = e => {
      // Enter
      if (e.key === "Enter" || e.key === " ") {
        onClick(e as any);
      }
    };
  }

  return <div {...innerProps}>{children}</div>;
};

export default Card;
