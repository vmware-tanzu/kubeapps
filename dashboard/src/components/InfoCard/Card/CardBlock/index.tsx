// Copyright 2020-2023 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import CardBlock from "./CardBlock";
export default CardBlock;
