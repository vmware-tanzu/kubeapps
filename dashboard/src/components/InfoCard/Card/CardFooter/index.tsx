// Copyright 2020-2023 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import CardFooter from "./CardFooter";
export default CardFooter;
