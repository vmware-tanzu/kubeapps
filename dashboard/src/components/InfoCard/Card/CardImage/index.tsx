// Copyright 2020-2023 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import CardImage from "./CardImage";
export default CardImage;
