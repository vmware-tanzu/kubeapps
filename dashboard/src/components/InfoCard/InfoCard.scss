// Copyright 2018-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

.card-header {
  .info-card-header {
    display: flex;
    justify-content: space-between;

    .card-title {
      max-width: calc(100% - 24px);
      margin-bottom: 0;
      overflow-x: hidden;
      text-overflow: ellipsis;
    }

    .card-tooltip {
      align-self: flex-end;
    }
  }
}

.bg-img {
  position: relative;
  right: 80px;
  bottom: 17px;
  width: 0;
  opacity: 0.05;

  img {
    height: 100px;
  }
}

.card-block {
  overflow: hidden;
  height: 4.5rem;

  .info-card-block {
    display: flex;

    .card-icon {
      display: flex;
      width: 64px;
      height: 64px;
      align-items: center;
      text-align: center;

      img {
        width: 64px;
        max-height: 64px;
      }
    }

    .card-description-wrapper {
      display: grid;
      width: 100%;
      height: 64px;
      align-items: center;

      .card-description {
        max-height: 64px;
        padding-left: 0.6rem;
      }
    }
  }
}

.kubeapps-card-footer {
  display: flex;
  width: 100%;
  justify-content: space-between;
  overflow-x: auto;

  .kubeapps-card-footer-col1 {
    margin: 0 0.6rem;
  }

  .kubeapps-card-footer-col2 {
    margin: 0 0.5rem 0 0;
  }
}

.card.clickable a {
  color: var(--cds-global-typography-color-400, #333);
  text-decoration: none;
}

.footer-tags {
  text-align: right;
}

.label.label-info-secondary {
  border-color: var(--cds-alias-status-neutral-shade, #00567a);
  background: var(--cds-alias-status-neutral-tint, #e3f5fc);
}
