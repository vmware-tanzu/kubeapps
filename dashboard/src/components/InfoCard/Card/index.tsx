// Copyright 2020-2023 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import Card from "./Card";
export default Card;
