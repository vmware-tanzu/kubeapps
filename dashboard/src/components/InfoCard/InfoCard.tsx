// Copyright 2018-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import Icon from "components/Icon/Icon";
import { Link } from "react-router-dom";
import placeholder from "icons/placeholder.svg";
import Card, { CardBlock, CardFooter, CardHeader } from "../js/Card";
import Column from "../js/Column";
import Row from "../js/Row";
import "./InfoCard.css";

export interface IInfoCardProps {
  title: string;
  info: string | JSX.Element;
  link?: string;
  icon?: string;
  bgIcon?: string;
  description?: string | JSX.Element;
  tooltip?: JSX.Element;
  tag1Class?: string;
  tag1Content?: string | JSX.Element;
  tag2Class?: string;
  tag2Content?: string | JSX.Element;
}

function InfoCard(props: IInfoCardProps) {
  const {
    title,
    link,
    info,
    description,
    tooltip,
    tag1Content,
    tag1Class,
    tag2Content,
    tag2Class,
    bgIcon,
  } = props;
  const icon = props.icon ? props.icon : placeholder;
  return (
    <Column span={[12, 6, 4, 3]}>
      <Card clickable={true}>
        <Link to={link || "#"}>
          <CardHeader>
            <div className="info-card-header">
              <div className="card-title">{title}</div>
              {tooltip ? <div className="card-tooltip">{tooltip}</div> : <></>}
            </div>
          </CardHeader>
          <CardBlock>
            <div className="info-card-block">
              <div className="card-icon">
                <Icon icon={icon} />
              </div>
              <div className="card-description-wrapper">
                <div className="card-description">
                  <span>{description}</span>
                </div>
              </div>
              {bgIcon ? (
                <div className="bg-img">
                  <img src={bgIcon} alt="bg-img" />
                </div>
              ) : (
                <></>
              )}
            </div>
          </CardBlock>
          <CardFooter>
            <Row>
              <div className="kubeapps-card-footer">
                <div className="kubeapps-card-footer-col1">{info}</div>
                <div className="kubeapps-card-footer-col2">
                  <div className="footer-tags">
                    {tag1Content && (
                      <span className={`label ${tag1Class || "label-info"}`}>{tag1Content}</span>
                    )}
                    {tag2Content && (
                      <span className={`label ${tag2Class || "label-info"}`}>{tag2Content}</span>
                    )}
                  </div>
                </div>
              </div>
            </Row>
          </CardFooter>
        </Link>
      </Card>
    </Column>
  );
}

export default InfoCard;
