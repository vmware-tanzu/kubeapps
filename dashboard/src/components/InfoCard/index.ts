// Copyright 2018-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import InfoCard from "./InfoCard";

export default InfoCard;
