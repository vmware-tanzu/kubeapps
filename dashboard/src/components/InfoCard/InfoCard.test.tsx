// Copyright 2018-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import Icon from "components/Icon/Icon";
import { shallow } from "enzyme";
import { Link } from "react-router-dom";
import { CardBlock } from "../js/Card";
import InfoCard from "./InfoCard";

it("should render a Card", () => {
  const wrapper = shallow(
    <InfoCard
      title="foo"
      info="foobar"
      link="/a/link/somewhere"
      icon="an-icon.png"
      tag1Class="blue"
      tag1Content="database"
      tag2Class="red"
      tag2Content="running"
    />,
  );
  expect(wrapper).toMatchSnapshot();
});

it("should generate a stub link if it's not provided", () => {
  const wrapper = shallow(
    <InfoCard
      title="foo"
      info="foobar"
      icon="an-icon.png"
      tag1Class="blue"
      tag1Content="database"
      tag2Class="red"
      tag2Content="running"
    />,
  );

  expect(wrapper.find(Link).prop("to")).toBe("#");
});

it("should avoid tags if they are not defined", () => {
  const wrapper = shallow(
    <InfoCard title="foo" info="foobar" link="/a/link/somewhere" icon="an-icon.png" />,
  );
  expect(wrapper.find(".ListItem__content__info_tag")).not.toExist();
});

it("should parse JSX elements in the tags", () => {
  const tag1 = <span className="tag1">tag1</span>;
  const tag2 = <span className="tag2">tag2</span>;
  const wrapper = shallow(
    <InfoCard
      title="foo"
      info="foobar"
      link="/a/link/somewhere"
      icon="an-icon.png"
      tag1Class="blue"
      tag1Content={tag1}
      tag2Class="red"
      tag2Content={tag2}
    />,
  );
  const tag1Elem = wrapper.find(".tag1");
  expect(tag1Elem).toExist();
  expect(tag1Elem.text()).toBe("tag1");
  const tag2Elem = wrapper.find(".tag2");
  expect(tag2Elem).toExist();
  expect(tag2Elem.text()).toBe("tag2");
});

it("should parse a description as text", () => {
  const wrapper = shallow(<InfoCard title="foo" info="foobar" description="a description" />);
  expect(wrapper.find(CardBlock).html()).toContain("a description");
});

it("should parse a description as JSX.Element", () => {
  const desc = <div className="description">This is a description</div>;
  const wrapper = shallow(<InfoCard title="foo" info="foobar" description={desc} />);
  expect(wrapper.find(".description").text()).toBe("This is a description");
});

it("should parse an icon", () => {
  const wrapper = shallow(<InfoCard title="foo" info="foobar" />);
  expect(wrapper.find(Icon)).toExist();
});

it("should parse a background img", () => {
  const wrapper = shallow(<InfoCard title="foo" info="foobar" bgIcon="img.png" />);
  expect(wrapper.find(".bg-img").find("img")).toExist();
});

it("should parse a tooltip component", () => {
  const wrapper = shallow(<InfoCard title="foo" info="foobar" tooltip={<div id="foo" />} />);
  expect(wrapper.find(".info-card-header").find("#foo")).toExist();
});
