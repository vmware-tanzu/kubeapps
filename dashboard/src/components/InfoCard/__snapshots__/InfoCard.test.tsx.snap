// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a Card 1`] = `
<Column
  listItem={false}
  span={
    Array [
      12,
      6,
      4,
      3,
    ]
  }
>
  <Card
    clickable={true}
    htmlTag="div"
  >
    <Link
      to="/a/link/somewhere"
    >
      <CardHeader
        noBorder={false}
      >
        <div
          className="info-card-header"
        >
          <div
            className="card-title"
          >
            foo
          </div>
        </div>
      </CardHeader>
      <CardBlock>
        <div
          className="info-card-block"
        >
          <div
            className="card-icon"
          >
            <Icon
              icon="an-icon.png"
            />
          </div>
          <div
            className="card-description-wrapper"
          >
            <div
              className="card-description"
            >
              <span />
            </div>
          </div>
        </div>
      </CardBlock>
      <CardFooter>
        <Row
          list={false}
        >
          <div
            className="kubeapps-card-footer"
          >
            <div
              className="kubeapps-card-footer-col1"
            >
              foobar
            </div>
            <div
              className="kubeapps-card-footer-col2"
            >
              <div
                className="footer-tags"
              >
                <span
                  className="label blue"
                >
                  database
                </span>
                <span
                  className="label red"
                >
                  running
                </span>
              </div>
            </div>
          </div>
        </Row>
      </CardFooter>
    </Link>
  </Card>
</Column>
`;
